File: src/Billing/UserAdminService.php
<?php
declare(strict_types=1);
namespace App\Billing;
use App\Infrastructure\Database;
use App\Integration\Provisioner;
final class UserAdminService
{
    public function __construct(private Database $db, private Wallet $wallet, private ?Provisioner $provisioner = null, private ?\App\Infrastructure\Outbox $outbox = null, private ?CustomerTimeline $timeline = null) {}
    /** Full user profile for the admin: account, balance, subscriptions, orders, promocodes, referrals. */
    public function profile(string $userId): array
    {
        $user = $this->db->one('SELECT * FROM users WHERE id=?', [$userId]);
        if (!$user) throw new BillingError('Пользователь не найден.');
        $subscriptions = $this->db->all(
            "SELECT s.*,COALESCE(o.plan_name,p.name) AS plan_name FROM subscriptions s LEFT JOIN orders o ON o.id=s.order_id LEFT JOIN plans p ON p.id=s.plan_id WHERE s.user_id=? ORDER BY s.created_at DESC",
            [$userId]
        );
        $orders = $this->db->all('SELECT * FROM orders WHERE user_id=? ORDER BY created_at DESC LIMIT 50', [$userId]);
        $topups = $this->db->all('SELECT * FROM topups WHERE user_id=? ORDER BY created_at DESC LIMIT 20', [$userId]);
        $transactions = $this->db->all('SELECT * FROM transactions WHERE user_id=? ORDER BY seq DESC LIMIT 30', [$userId]);
        $promoUses = $this->db->all(
            'SELECT pu.used_at,p.code,p.type,p.balance_bonus_kopeks,p.subscription_days FROM promocode_uses pu JOIN promocodes p ON p.id=pu.promocode_id WHERE pu.user_id=? ORDER BY pu.used_at DESC',
            [$userId]
        );
        $referrals = $this->db->all('SELECT id,email,telegram_id,created_at,has_made_first_topup FROM users WHERE referred_by_id=? ORDER BY created_at DESC', [$userId]);
        $referrer = $user['referred_by_id'] ? $this->db->one('SELECT id,email,telegram_id FROM users WHERE id=?', [$user['referred_by_id']]) : null;
        $spending = $this->db->one("SELECT COALESCE(SUM(amount_kopeks),0) AS s FROM transactions WHERE user_id=? AND amount_kopeks<0 AND type IN ('subscription_purchase','subscription_renewal','gift_purchase','traffic_topup','device_addon')", [$userId]);
        $gifts = $this->db->all('SELECT * FROM guest_purchases WHERE buyer_user_id=? AND is_gift=1 ORDER BY created_at DESC', [$userId]);
        return [
            'user' => $user,
            'subscriptions' => $subscriptions,
            'orders' => $orders,
            'topups' => $topups,
            'transactions' => $transactions,
            'promo_uses' => $promoUses,
            'referrals' => $referrals,
            'referrer' => $referrer,
            'spent_kopeks' => -(int)($spending['s'] ?? 0),
            'gifts' => $gifts,
            'timeline' => $this->timeline?->forUser($userId) ?? [],
        ];
    }
    /** Adjust user balance (credit or debit) with audit. */
    public function adjustBalance(string $userId, int $amountKopeks, string $reason, string $actor): void
    {
        if ($amountKopeks === 0) throw new BillingError('Сумма не может быть нулевой.');
        if (abs($amountKopeks) > 100000000) throw new BillingError('Сумма слишком большая.');
        if (mb_strlen($reason) < 3 || mb_strlen($reason) > 200) throw new BillingError('Причина: 3–200 символов.');
        $this->db->transaction(function () use ($userId, $amountKopeks, $reason, $actor) {
            if ($amountKopeks > 0) {
                $this->wallet->credit($userId, $amountKopeks, 'manual_adjust', 'Ручная корректировка: '.$reason);
            } else {
                $this->wallet->debit($userId, -$amountKopeks, 'manual_adjust', 'Ручная корректировка: '.$reason);
            }
            $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $actor, 'user.balance_adjusted', $userId, time()]);
        });
    }
    /** Grant subscription days to a user (extend or create). */
    public function grantDays(string $userId, int $days, ?string $planId, string $actor): array
    {
        if ($days < 1 || $days > 3650) throw new BillingError('Дни: 1–3650.');
        return $this->db->transaction(function () use ($userId, $days, $planId, $actor) {
            $sub = $this->db->one("SELECT * FROM subscriptions WHERE user_id=? AND status IN ('active','trial','provisioning') ORDER BY created_at DESC LIMIT 1" . $this->db->lock(), [$userId]);
            if ($sub) {
                $base = max(time(), (int)$sub['expires_at']);
                $this->db->execute("UPDATE subscriptions SET expires_at=?,status='active',updated_at=? WHERE id=?", [$base + $days * 86400, time(), $sub['id']]);
                $this->outbox?->enqueue('subscription.extend','admin-extend:'.Database::id(),['subscription_id'=>$sub['id']]);
                $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $actor, 'user.days_granted', $userId, time()]);
                return $this->db->one('SELECT * FROM subscriptions WHERE id=?', [$sub['id']]);
            }
            $plan = $planId ? $this->db->one('SELECT * FROM plans WHERE id=?', [$planId]) : null;
            $id = Database::id();
            $now = time();
            $this->db->execute(
                "INSERT INTO subscriptions(id,order_id,user_id,status,expires_at,created_at,plan_id,traffic_limit_gb,device_limit,is_trial,start_date,updated_at) VALUES(?,NULL,?,'active',?,?,?,?,?,0,?,?)",
                [$id, $userId, $now + $days * 86400, $now, $plan['id'] ?? null, $plan ? (int)$plan['traffic_bytes'] / 1073741824 : 0, $plan ? (int)$plan['devices'] : 1, $now, $now]
            );
            $this->outbox?->enqueue('subscription.provision','provision:'.$id,['subscription_id'=>$id]);
            $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $actor, 'user.days_granted', $userId, time()]);
            return $this->db->one('SELECT * FROM subscriptions WHERE id=?', [$id]);
        });
    }
    /** Set a promo discount for a user. */
    public function setDiscount(string $userId, int $percent, int $hours, string $actor): void
    {
        if ($percent < 1 || $percent > 99) throw new BillingError('Скидка: 1–99%.');
        if ($hours < 0 || $hours > 8760) throw new BillingError('Часы: 0–8760.');
        $this->db->execute('UPDATE users SET promo_offer_discount_percent=?,promo_offer_discount_source=?,promo_offer_discount_expires_at=? WHERE id=?', [$percent, 'admin:'.$actor, $hours > 0 ? time() + $hours * 3600 : null, $userId]);
        $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $actor, 'user.discount_set', $userId, time()]);
    }
    public function clearDiscount(string $userId, string $actor): void
    {
        $this->db->execute('UPDATE users SET promo_offer_discount_percent=0,promo_offer_discount_source=NULL,promo_offer_discount_expires_at=NULL WHERE id=?', [$userId]);
        $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $actor, 'user.discount_cleared', $userId, time()]);
    }
    /** Full subscription view with live panel data (traffic used, devices, status). */
    public function subscription(string $subscriptionId): array
    {
        $sub = $this->db->one(
            "SELECT s.*,COALESCE(o.plan_name,p.name) AS plan_name,u.email,u.telegram_id FROM subscriptions s LEFT JOIN orders o ON o.id=s.order_id LEFT JOIN plans p ON p.id=s.plan_id LEFT JOIN users u ON u.id=s.user_id WHERE s.id=?",
            [$subscriptionId]
        );
        if (!$sub) throw new BillingError('Подписка не найдена.');
        $panel = null;
        if ($this->provisioner) {
            try {
                if ((int)($sub['remnawave_id'] ?? 0) > 0) {
                    $panel = $this->provisioner->fetchById((int)$sub['remnawave_id']);
                } else {
                    $panel = $this->provisioner->fetch('zb_'.$sub['id']);
                }
            } catch (\Throwable $e) {}
        }
        $panelTrafficUsed = 0;
        if ($panel) {
            $panelTrafficUsed = (int)($panel['userTraffic']['usedTrafficBytes'] ?? 0) / 1073741824;
        }
        return [
            'sub' => $sub,
            'panel' => $panel,
            'panel_traffic_used_gb' => (float)$panelTrafficUsed,
            'plans' => $this->db->all('SELECT id,name,duration_days,traffic_bytes,devices FROM plans ORDER BY display_order'),
        ];
    }
    /** Update subscription traffic limit (GB) locally + push to Remnawave. */
    public function updateSubscriptionTraffic(string $userId, string $subscriptionId, int $trafficGb, string $actor): void
    {
        if ($trafficGb < 0 || $trafficGb > 100000) throw new BillingError('Трафик 0–100000 ГБ.');
        $this->db->transaction(function () use ($userId, $subscriptionId, $trafficGb, $actor) {
            $sub = $this->db->one('SELECT * FROM subscriptions WHERE id=?' . $this->db->lock(), [$subscriptionId]);
            if (!$sub || $sub['user_id'] !== $userId) throw new BillingError('Подписка не найдена.');
            $this->db->execute('UPDATE subscriptions SET traffic_limit_gb=? WHERE id=?', [$trafficGb, $subscriptionId]);
            $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $actor, 'user.subscription_traffic_changed', $userId, time()]);
        });
        $this->applyPanelChanges($subscriptionId);
    }
    /** Update subscription device limit locally + push to Remnawave (0 = unlimited). */
    public function updateSubscriptionDevices(string $userId, string $subscriptionId, int $devices, string $actor): void
    {
        if ($devices < 0 || $devices > 20) throw new BillingError('Устройства 0–20 (0 = безлимит).');
        $this->db->transaction(function () use ($userId, $subscriptionId, $devices, $actor) {
            $sub = $this->db->one('SELECT * FROM subscriptions WHERE id=?' . $this->db->lock(), [$subscriptionId]);
            if (!$sub || $sub['user_id'] !== $userId) throw new BillingError('Подписка не найдена.');
            $this->db->execute('UPDATE subscriptions SET device_limit=? WHERE id=?', [$devices, $subscriptionId]);
            $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $actor, 'user.subscription_devices_changed', $userId, time()]);
        });
        $this->applyPanelChanges($subscriptionId);
    }
    /** Extend subscription by N days (from now or from current expiry), sync panel expiry. */
    public function extendSubscription(string $userId, string $subscriptionId, int $days, string $actor): void
    {
        if ($days < 1 || $days > 3650) throw new BillingError('Дни: 1–3650.');
        $this->db->transaction(function () use ($userId, $subscriptionId, $days, $actor) {
            $sub = $this->db->one('SELECT * FROM subscriptions WHERE id=?' . $this->db->lock(), [$subscriptionId]);
            if (!$sub || $sub['user_id'] !== $userId) throw new BillingError('Подписка не найдена.');
            $base = max(time(), (int)$sub['expires_at']);
            $this->db->execute("UPDATE subscriptions SET expires_at=?,status='active',updated_at=? WHERE id=?", [$base + $days * 86400, time(), $subscriptionId]);
            $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $actor, 'user.subscription_extended', $userId, time()]);
        });
        $this->applyPanelChanges($subscriptionId);
    }
    /** Set an exact expiration datetime (UTC timestamp) for one subscription. */
    public function setSubscriptionExpiry(string $userId, string $subscriptionId, int $expiresAt, string $actor): void
    {
        $min = time() - 3600; // allow slightly-past; anything older is a mistake
        if ($expiresAt < $min) throw new BillingError('Дата окончания не может быть в прошлом.');
        if ($expiresAt > time() + 3650 * 86400) throw new BillingError('Дата окончания слишком далеко (макс. 10 лет).');
        $this->db->transaction(function () use ($userId, $subscriptionId, $expiresAt, $actor) {
            $sub = $this->db->one('SELECT * FROM subscriptions WHERE id=?' . $this->db->lock(), [$subscriptionId]);
            if (!$sub || $sub['user_id'] !== $userId) throw new BillingError('Подписка не найдена.');
            $this->db->execute("UPDATE subscriptions SET expires_at=?,status='active',updated_at=? WHERE id=?", [$expiresAt, time(), $subscriptionId]);
            $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $actor, 'user.subscription_expiry_set', $userId, time()]);
        });
        $this->applyPanelChanges($subscriptionId);
    }
    /** Reset used traffic on panel (bytesUsed back to 0). */
    public function resetSubscriptionTraffic(string $userId, string $subscriptionId, string $actor): void
    {
        $this->db->transaction(function () use ($userId, $subscriptionId, $actor) {
            $sub = $this->db->one('SELECT * FROM subscriptions WHERE id=?' . $this->db->lock(), [$subscriptionId]);
            if (!$sub || $sub['user_id'] !== $userId) throw new BillingError('Подписка не найдена.');
            $this->db->execute('UPDATE subscriptions SET traffic_used_gb=0,updated_at=? WHERE id=?', [time(), $subscriptionId]);
            $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $actor, 'user.subscription_traffic_reset', $userId, time()]);
        });
        $this->applyPanelChanges($subscriptionId);
    }
    /** Push current local values (traffic/devices/expiry) to Remnawave panel. */
    private function applyPanelChanges(string $subscriptionId): void
    {
        if (!$this->provisioner) return;
        $sub = $this->db->one('SELECT * FROM subscriptions WHERE id=?', [$subscriptionId]);
        if (!$sub || $sub['status']!=='active') return;
        $sub['traffic_bytes'] = (int)$sub['traffic_limit_gb'] * 1073741824;
        $sub['devices'] = (int)$sub['device_limit'];
        try {
            $panel = null;
            if ((int)($sub['remnawave_id'] ?? 0) > 0) {
                $panel = $this->provisioner->fetchById((int)$sub['remnawave_id']);
                if ($panel && ($panel['id'] ?? null)) {
                    $this->provisioner->updateById((int)$panel['id'], (int)$sub['traffic_bytes'], (int)$sub['devices'], (int)$sub['expires_at']);
                    return;
                }
            }
            $panel = $this->provisioner->fetch('zb_'.$sub['id']);
            if ($panel && ($panel['id'] ?? null)) {
                $this->provisioner->extend($sub); // PATCHes traffic + devices + expireAt by username
            }
        } catch (\Throwable $e) {}
    }
    /** Remove a subscription: deletes the panel user (Remnawave) and the local row. */
    public function removeSubscription(string $userId, string $subscriptionId, string $actor): void
    {
        $this->db->transaction(function () use ($userId, $subscriptionId, $actor) {
            $sub = $this->db->one('SELECT * FROM subscriptions WHERE id=?' . $this->db->lock(), [$subscriptionId]);
            if (!$sub || $sub['user_id'] !== $userId) throw new BillingError('Подписка не найдена.');
            // 1) Delete the VPN user from Remnawave (legacy subs use panel id, new ones use zb_<id>).
            if ($this->provisioner) {
                $panelId = (int)($sub['remnawave_id'] ?? 0);
                if ($panelId > 0) {
                    if (method_exists($this->provisioner, 'removeById')) {
                        $this->provisioner->removeById($panelId);
                    } else {
                        $this->provisioner->remove('zb_' . $sub['id']);
                    }
                } else {
                    $this->provisioner->remove('zb_' . $sub['id']);
                }
            }
            // 2) Drop pending ops for this subscription.
            $this->db->execute("DELETE FROM outbox WHERE topic IN ('subscription.provision','subscription.extend') AND payload LIKE ?", ['%'.$subscriptionId.'%']);
            // 3) Delete local row (no FK constraints on subscriptions).
            $this->db->execute('DELETE FROM subscriptions WHERE id=?', [$subscriptionId]);
            $this->db->execute('INSERT INTO audit_log VALUES(?,?,?,?,?)', [Database::id(), $actor, 'user.subscription_removed', $userId, time()]);
        });
    }
    /** Search users by email, telegram id, referral code. */
    public function search(string $query, int $limit = 50): array
    {
        $q = trim($query);
        if ($q === '') return [];
        $like = '%'.$q.'%';
        return $this->db->all(
            'SELECT id,email,telegram_id,role,disabled,balance_kopeks,created_at FROM users WHERE email LIKE ? OR telegram_id LIKE ? OR referral_code LIKE ? ORDER BY created_at DESC LIMIT ?',
            [$like, $like, $like, $limit]
        );
    }
}

File: rails_backend/app/services/admin/user_admin_service.rb
module Admin
  # Admin operations on customer accounts and subscriptions. All local changes,
  # wallet ledger entries, audit records, and outbox messages commit together.
  class UserAdminService
    def initialize(wallet: Billing::WalletService.new, outbox: Infrastructure::OutboxService.new)
      @wallet = wallet
      @outbox = outbox
    end

    def profile(user_id)
      user = User.find_by(id: user_id)
      raise Billing::BillingError, "Пользователь не найден." unless user

      promo_uses = PromocodeUse.joins(:promocode).where(user_id: user.id)
        .order(used_at: :desc).map do |use|
          use.attributes.slice("used_at").merge(use.promocode.attributes.slice(
            "code", "type", "balance_bonus_kopeks", "subscription_days"))
        end
      spent = TransactionRecord.where(user_id: user.id, type: %w[
        subscription_purchase subscription_renewal gift_purchase traffic_topup device_addon
      ]).where("amount_kopeks < 0").sum(:amount_kopeks)

      {
        user: user,
        subscriptions: Subscription.where(user_id: user.id).order(created_at: :desc).map do |sub|
          sub.as_json.merge("plan_name" => sub.order&.plan_name || Plan.find_by(id: sub.plan_id)&.name)
        end,
        orders: Order.where(user_id: user.id).order(created_at: :desc).limit(50),
        topups: Topup.where(user_id: user.id).order(created_at: :desc).limit(20),
        transactions: TransactionRecord.where(user_id: user.id).order(seq: :desc).limit(30),
        promo_uses: promo_uses,
        referrals: User.where(referred_by_id: user.id).order(created_at: :desc)
          .as_json(only: %i[id email telegram_id created_at has_made_first_topup]),
        referrer: user.referred_by_id.present? ? User.where(id: user.referred_by_id)
          .as_json(only: %i[id email telegram_id]).first : nil,
        spent_kopeks: -spent.to_i,
        gifts: GuestPurchase.where(buyer_user_id: user.id, is_gift: 1).order(created_at: :desc),
        timeline: CustomerTimeline.where(user_id: user.id).order(recorded_at: :asc).limit(100)
          .map { |event| event.as_json(only: %i[event_type payload occurred_at]) }
      }
    end

    def adjust_balance(user_id:, amount_kopeks:, reason:, actor:)
      amount = integer!(amount_kopeks, "Сумма не может быть нулевой.")
      reason = reason.to_s.strip
      raise Billing::BillingError, "Сумма не может быть нулевой." if amount.zero?
      raise Billing::BillingError, "Сумма слишком большая." if amount.abs > 100_000_000
      raise Billing::BillingError, "Причина: 3–200 символов." unless reason.length.between?(3, 200)

      ApplicationRecord.transaction do
        user_exists!(user_id)
        if amount.positive?
          @wallet.credit(user_id, amount, "manual_adjust", "Ручная корректировка: #{reason}")
        else
          @wallet.debit(user_id, -amount, "manual_adjust", "Ручная корректировка: #{reason}")
        end
        audit(actor, "user.balance_adjusted", user_id)
      end
      true
    end

    def grant_days(user_id:, days:, plan_id: nil, actor:)
      days = bounded_integer!(days, 1..3650, "Дни: 1–3650.")
      ApplicationRecord.transaction do
        user = user_exists!(user_id)
        sub = Subscription.lock.where(user_id: user.id, status: %w[active trial provisioning])
          .order(created_at: :desc).first
        if sub
          base = [Time.now.to_i, sub.expires_at.to_i].max
          provisioning = sub.status == "provisioning"
          sub.update!(expires_at: base + days * 86_400,
            status: provisioning ? "provisioning" : "active", updated_at: Time.now.to_i)
          enqueue_subscription(provisioning ? "subscription.provision" : "subscription.extend", sub.id,
            "admin-grant:#{Infrastructure::IdGenerator.call}")
        else
          plan = plan_id.present? ? Plan.active.find_by(id: plan_id) : nil
          raise Billing::BillingError, "Для новой подписки выберите активный тариф." unless plan
          now = Time.now.to_i
          sub = Subscription.create!(id: Infrastructure::IdGenerator.call, order_id: nil,
            user_id: user.id, status: "provisioning", expires_at: now + days * 86_400,
            created_at: now, plan_id: plan.id,
            traffic_limit_gb: plan.traffic_bytes.to_i / 1.gigabyte,
            device_limit: plan.devices, is_trial: 0, start_date: now, updated_at: now)
          enqueue_subscription("subscription.provision", sub.id, "provision:#{sub.id}")
        end
        audit(actor, "user.days_granted", user.id)
        sub
      end
    end

    def set_discount(user_id:, percent:, hours:, actor:)
      percent = bounded_integer!(percent, 1..99, "Скидка: 1–99%.")
      hours = bounded_integer!(hours, 0..8760, "Часы: 0–8760.")
      ApplicationRecord.transaction do
        user = user_exists!(user_id)
        user.update!(promo_offer_discount_percent: percent,
          promo_offer_discount_source: "admin:#{actor}",
          promo_offer_discount_expires_at: hours.positive? ? Time.now.to_i + hours * 3600 : nil)
        audit(actor, "user.discount_set", user.id)
      end
      true
    end

    def clear_discount(user_id:, actor:)
      ApplicationRecord.transaction do
        user = user_exists!(user_id)
        user.update!(promo_offer_discount_percent: 0, promo_offer_discount_source: nil,
          promo_offer_discount_expires_at: nil)
        audit(actor, "user.discount_cleared", user.id)
      end
      true
    end

    def subscription(subscription_id)
      sub = Subscription.includes(:order, :user).find_by(id: subscription_id)
      raise Billing::BillingError, "Подписка не найдена." unless sub

      {
        sub: sub.as_json.merge("plan_name" => sub.order&.plan_name || Plan.find_by(id: sub.plan_id)&.name,
          "email" => sub.user.email, "telegram_id" => sub.user.telegram_id),
        panel: ProvisioningAccount.find_by(subscription_id: sub.id)&.as_json,
        panel_traffic_used_gb: sub.traffic_used_gb.to_f,
        plans: Plan.order(:display_order, :name).as_json(only: %i[id name duration_days traffic_bytes devices])
      }
    end

    def update_subscription_traffic(user_id:, subscription_id:, traffic_gb:, actor:)
      value = bounded_integer!(traffic_gb, 0..100_000, "Трафик 0–100000 ГБ.")
      update_owned_subscription(user_id, subscription_id, actor, "user.subscription_traffic_changed") do |sub|
        sub.update!(traffic_limit_gb: value, updated_at: Time.now.to_i)
      end
    end

    def update_subscription_devices(user_id:, subscription_id:, devices:, actor:)
      value = bounded_integer!(devices, 0..20, "Устройства 0–20 (0 = безлимит).")
      update_owned_subscription(user_id, subscription_id, actor, "user.subscription_devices_changed") do |sub|
        sub.update!(device_limit: value, updated_at: Time.now.to_i)
      end
    end

    def extend_subscription(user_id:, subscription_id:, days:, actor:)
      value = bounded_integer!(days, 1..3650, "Дни: 1–3650.")
      update_owned_subscription(user_id, subscription_id, actor, "user.subscription_extended") do |sub|
        base = [Time.now.to_i, sub.expires_at.to_i].max
        sub.update!(expires_at: base + value * 86_400, status: "active", updated_at: Time.now.to_i)
      end
    end

    def set_subscription_expiry(user_id:, subscription_id:, expires_at:, actor:)
      value = integer!(expires_at, "Укажите дату окончания правильно.")
      now = Time.now.to_i
      raise Billing::BillingError, "Дата окончания не может быть в прошлом." if value < now - 3600
      raise Billing::BillingError, "Дата окончания слишком далеко (макс. 10 лет)." if value > now + 3650 * 86_400
      update_owned_subscription(user_id, subscription_id, actor, "user.subscription_expiry_set") do |sub|
        sub.update!(expires_at: value, status: "active", updated_at: now)
      end
    end

    def reset_subscription_traffic(user_id:, subscription_id:, actor:)
      update_owned_subscription(user_id, subscription_id, actor, "user.subscription_traffic_reset") do |sub|
        sub.update!(traffic_used_gb: 0, updated_at: Time.now.to_i)
      end
    end

    # Mark the row disabled and enqueue remote removal. Keeping the local record
    # preserves order/payment references and avoids a network request inside a DB transaction.
    def remove_subscription(user_id:, subscription_id:, actor:)
      ApplicationRecord.transaction do
        sub = owned_subscription!(user_id, subscription_id)
        sub.update!(status: "disabled", lifecycle_status: "cancelled", updated_at: Time.now.to_i)
        OutboxJob.where(topic: %w[
          subscription.provision subscription.extend subscription.renew subscription.daily
          subscription.traffic subscription.devices subscription.admin_sync
        ], status: "pending").find_each do |job|
          next unless JSON.parse(job.payload.to_s).fetch("subscription_id", nil).to_s == subscription_id.to_s
          job.update!(status: "done", payload: "{}")
        rescue JSON::ParserError
          # Leave malformed jobs to the worker's existing poison-message policy.
        end
        @outbox.enqueue("subscription.remove", "admin-remove:#{subscription_id}:#{Infrastructure::IdGenerator.call}",
          { subscription_id: subscription_id, requested_by: actor.to_s })
        audit(actor, "user.subscription_removed", user_id)
      end
      true
    end

    def search(query, limit: 50)
      q = query.to_s.strip
      return [] if q.empty?
      limit = [[Integer(limit), 1].max, 200].min
      pattern = "%#{ActiveRecord::Base.sanitize_sql_like(q)}%"
      User.where("LOWER(COALESCE(email, '')) LIKE LOWER(:q) OR LOWER(COALESCE(telegram_id, '')) LIKE LOWER(:q) OR LOWER(COALESCE(referral_code, '')) LIKE LOWER(:q)", q: pattern)
        .order(created_at: :desc).limit(limit)
        .as_json(only: %i[id email telegram_id role disabled balance_kopeks created_at])
    rescue ArgumentError, TypeError
      raise Billing::BillingError, "Некорректный лимит поиска."
    end

    private

    def update_owned_subscription(user_id, subscription_id, actor, action)
      ApplicationRecord.transaction do
        sub = owned_subscription!(user_id, subscription_id)
        yield sub
        enqueue_subscription("subscription.admin_sync", sub.id, "admin-sync:#{Infrastructure::IdGenerator.call}")
        audit(actor, action, user_id)
        sub
      end
    end

    def owned_subscription!(user_id, subscription_id)
      sub = Subscription.lock.find_by(id: subscription_id)
      raise Billing::BillingError, "Подписка не найдена." unless sub && sub.user_id == user_id.to_s
      sub
    end

    def user_exists!(user_id)
      User.lock.find_by(id: user_id).tap do |user|
        raise Billing::BillingError, "Пользователь не найден." unless user
      end
    end

    def enqueue_subscription(topic, subscription_id, key)
      @outbox.enqueue(topic, key, { subscription_id: subscription_id })
    end

    def audit(actor, action, subject)
      AuditLog.create!(id: Infrastructure::IdGenerator.call, actor: actor.to_s,
        action: action, subject: subject.to_s, created_at: Time.now.to_i)
    end

    def integer!(value, error)
      parsed = Integer(value, exception: false)
      raise Billing::BillingError, error if parsed.nil?
      parsed
    end

    def bounded_integer!(value, range, error)
      parsed = integer!(value, error)
      raise Billing::BillingError, error unless range.cover?(parsed)
      parsed
    end
  end
end
